src/Uno.UI/Microsoft/UI/Xaml/Controls/RatingControl/RatingControl.cs

```diff
@@ -0,0 +1,1205 @@
+// MUX Reference RatingControl.cpp, commit de78834
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using Uno.UI.Helpers.WinUI;
+using Windows.Foundation;
+using Windows.System;
+using Windows.UI.Composition;
+using Windows.UI.ViewManagement;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Hosting;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Automation.Peers;
+using System.Numerics;
+using Windows.Devices.Input;
+using RatingControlAutomationPeer = Microsoft.UI.Xaml.Automation.Peers.RatingControlAutomationPeer;
+
+namespace Microsoft.UI.Xaml.Controls
+{
+	/// <summary>
+	/// Represents a control that lets a user enter a star rating.
+	/// </summary>
+	public partial class RatingControl : Control
+	{
+#pragma warning disable IDE0051 // Remove unused private members
+		private const float c_horizontalScaleAnimationCenterPoint = 0.5f;
+#pragma warning restore IDE0051 // Remove unused private members
+		private const float c_verticalScaleAnimationCenterPoint = 0.8f;
+		private readonly Thickness c_focusVisualMargin = new Thickness(-8, -7, -8, 0);
+		private const int c_defaultRatingFontSizeForRendering = 32; // (32 = 2 * [default fontsize] -- because of double size rendering), remove when MSFT #10030063 is done
+		private const int c_defaultItemSpacing = 8;
+
+		private const float c_mouseOverScale = 0.8f;
+		private const float c_touchOverScale = 1.0f;
+		private const float c_noPointerOverMagicNumber = -100;
+
+		// 22 = 20(compensate for the -20 margin on StackPanel) + 2(magic number makes the text and star center-aligned)
+		private const float c_defaultCaptionTopMargin = 22;
+
+		private const int c_noValueSetSentinel = -1;
+
+		private static UISettings _uiSettings = null;
+
+		private DispatcherHelper m_dispatcherHelper;
+
+		// Private members
+		private TextBlock m_captionTextBlock;
+
+		private CompositionPropertySet m_sharedPointerPropertySet;
+
+		private StackPanel m_backgroundStackPanel;
+		private StackPanel m_foregroundStackPanel;
+
+		private bool m_isPointerOver = false;
+		private bool m_isPointerDown = false;
+		private double m_mousePercentage = 0.0;
+
+		private RatingInfoType m_infoType = RatingInfoType.Font;
+		private double m_preEngagementValue = 0.0;
+		private bool m_disengagedWithA = false;
+		private bool m_shouldDiscardValue = true;
+		private long m_fontFamilyChangedToken;
+
+		/// <summary>
+		/// Initializes a new instance of the RatingControl class.
+		/// </summary>
+		public RatingControl()
+		{
+			m_dispatcherHelper = new DispatcherHelper(this);
+			//__RP_Marker_ClassById(RuntimeProfiler.ProfId_RatingControl);
+
+			DefaultStyleKey = typeof(RatingControl);
+		}
+
+		~RatingControl()
+		{
+			// We only need to use safe_get in the deruction loop
+			RecycleEvents(true /* useSafeGet */);
```

Comment:
Probably not in case of the constructor, but it is likely needed in `OnApplyTemplate` in case template changes on the fly. There are no leak tests though...

Reply:
Unhooking in the destructor is actually a concurrency issue, particularly if it relates to DependencyObject (which assumes everything's done on the owning dispatcher). Performing this during Load/Unload is best.

Reply:
Good to know, will adjust it accrodingly